### PR TITLE
Check that abstract reference base objects are coercible to objects

### PR DESCRIPTION
**Pull request by hermanventer**
The base value of a Reference is an abstract value that is not known to be an object or one of a set of objects, then the effects of an assignment to the reference is unknown to the extent that Prepack cannot continue. It also had the effect that an invariant was violated, leading to issue #750.

We now give a error message that is contextualized to the member expression (if any) that created  the Reference.

[PP0012 reference base value is unknown](https://github.com/facebook/prepack/wiki/PP0012)

### CODE REVIEW

File: src/methods/abstract.js [hunk 1]
```diff
@@ -54,11 +56,23 @@ export function SplitMatch(realm: Realm, S: string, q: number, R: string): false
 }
 
 // ECMA262 7.2.1
-export function RequireObjectCoercible<T: Value>(realm: Realm, arg: T): T {
-  if (HasSomeCompatibleType(arg, NullValue, UndefinedValue)) {
+export function RequireObjectCoercible(
```
Comment (cblappert) [inline src/methods/abstract.js]: This typing seems less specific than the old typing, is it possible to leave it templated and have flow enforce that?

File: src/methods/environment.js [hunk 2]
```diff
@@ -77,7 +77,11 @@ export function IsSuperReference(realm: Realm, V: Reference): boolean {
 // HasPrimitiveBase(V). Returns true if Type(base) is Boolean, String, Symbol, or Number.
 export function HasPrimitiveBase(realm: Realm, V: Reference): boolean {
   let base = GetBase(realm, V);
-  return base instanceof Value && HasSomeCompatibleType(base, BooleanValue, StringValue, SymbolValue, NumberValue);
+  // void | ObjectValue | BooleanValue | StringValue | SymbolValue | NumberValue | EnvironmentRecord | AbstractValue;
+  if (!base || base instanceof EnvironmentRecord) return false;
+  if (base instanceof ObjectValue || base instanceof AbstractObjectValue) return false;
```
Comment (cblappert) [inline src/methods/environment.js]: Why is this needed as well as the checks in the return statement?

File: src/methods/abstract.js (hunk 1 above)
Comment (hermanventer) [inline src/methods/abstract.js]: The reason I changed this is that the template does not exclude the possibility of the result being an instance of NullValue or UndefinedValue. The added precision of the other cases is not actually needed anywhere.

File: src/methods/environment.js (hunk 2 above)
Comment (hermanventer) [inline src/methods/environment.js]: It is not needed, but calling them out explicitly makes it more obvious that the code is checking each case in the type union.
